VstsTfsSample/Program.cs

```diff
@@ -0,0 +1,122 @@
+using System.Collections.Generic;
+
+namespace VstsTfsSample
+{
+    class Program
+    {
+        //============= Config [Edit these with your settings] =====================
+        internal const string vstsOrTfsCollectionUrl = "https://myaccount.visualstudio.com"; //change to the URL of your VSTS or TFS account
+        internal const string clientId = "0fa17cf4-f75c-4185-ab9a-7c5ea47ca073"; //change to your app registration's Application ID (For VSTS ADAL authentication)
+        internal const string replyUri = "http://adalsample"; //change to your app registration's reply URI. (For VSTS ADAL authentication)
+        //==========================================================================
+
+        internal const string VSTSResourceId = "499b84ac-1321-427f-aa17-267ca6975798"; //Constant value to target VSTS. Do not change  
+
+        public static void Main(string[] args)
+        {
+            //Decipher VSTS or TFS collection
+            if (isVsts(vstsOrTfsCollectionUrl))
+            {
+                //VSTS ADAL authentication
+                AuthenticationContext ctx = GetAuthenticationContext(null);
+                AuthenticationResult result = null;
+                try
+                {
+                    //PromptBehavior.RefreshSession will enforce an authn prompt every time. NOTE: Auto will take your windows login state if possible
+                    result = ctx.AcquireTokenAsync(VSTSResourceId, clientId, new Uri(replyUri), new PlatformParameters(PromptBehavior.Always)).Result;
+                    Console.WriteLine("Token expires on: " + result.ExpiresOn);
+
+                    var bearerAuthHeader = new AuthenticationHeaderValue("Bearer", result.AccessToken);
+                    //using access token to call REST endpoint
+                    ListProjects(bearerAuthHeader);
+                }
+                catch (UnauthorizedAccessException)
+                {
+                    // If the token has expired, prompt the user with a login prompt
+                    result = ctx.AcquireTokenAsync(VSTSResourceId, clientId, new Uri(replyUri), new PlatformParameters(PromptBehavior.Always)).Result;
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine("{0}: {1}", ex.GetType(), ex.Message);
+                }
+            } else
+            {
+                //TFS Windows credential authentication
+                VssConnection connection = new VssConnection(new Uri(vstsOrTfsCollectionUrl), new VssClientCredentials());
+
+                ProjectHttpClient projectClient = connection.GetClient<ProjectHttpClient>();
+                IEnumerable<TeamProjectReference> projects = projectClient.GetProjects().Result;
+                foreach (TeamProjectReference p in projects)
+                {
+                    Console.WriteLine(p.Name);
+                }
+            }
+        }
+
+        private static bool isVsts(string collectionUrl)
```

Comment:
there HAS to be a better solution than this...  Let's talk to Will and see if he has any ideas.  This is likely to break when we change our URLs for Azure Front Door...